### PR TITLE
Add DataFrame.T and DataFrame.transpose

### LINKED ISSUE #467: Missing T property
As mentioned in #382, Koalas is currently missing the `DataFrame.T` property. Since I personally use that property a lot when working with pandas, I'd be happy if it was available in Koalas as well. However, because of its distributed nature, simply turning rows into columns does not easily work with Spark. After thinking about how to implement `T` with Spark for a while now, I couldn't find any solution that is feasible for large datasets. Does anyone have an idea how to actually implement this or should we just raise a `NotImplementedError` that explains why `T` isn't supported by Koalas?

### PR DESCRIPTION
**Pull request by HyukjinKwon**
This PR adds DataFrame.T and DataFrame.transpose.

This will be an expensive operation ... but still work.

Resolves #467

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @rxin, @thunterdb,  @ueshin, @floscha, @garawalid

**Comment by HyukjinKwon** [comment]
What do you guys think about this?

**Comment by ueshin** [comment]
The implementation itself LGTM, but I'd say this is too dangerous for big data.
At least, how about adding the limitation to support number of rows, or adding an argument or a config to limit the rows for users to configure? Maybe we need a note in documents as well.

**Comment by HyukjinKwon** [comment]
Let me at least leave a note like some other APIs we have (e.g., to_pandas)

**Comment by HyukjinKwon** [comment]
Tests passed

**Comment by HyukjinKwon** [comment]
@rxin, @shril, @floscha, @ueshin, if you guys don't mind, I will add the limitation logic later.
To check the length, the only idea I have it is to check the length of each partition.

To do this, we should call `mapPartitions` which is expensive. I think it's a bit odd that we add an expensive operation to check the length.

I am checking the possibility to have an API like `mapPartitionsInArrow`. Maybe, I will integrate it later if you guys don't mind.

**Comment by HyukjinKwon** [comment]
Can I merge this one first, @rxin if you dont mind? or you want to add the length checking together?

**Comment by rxin** [comment]
I'd add a simple length check, e.g. we should not have more than 1000 rows.

**Comment by HyukjinKwon** [comment]
@rxin, sorry but could I ask kind of a simple length check do you mean?

- `mapPartitions` : it needs to serialize and deserialize whole data row by row.
- `take(1000)`: it could lead execute twice the input dataframe.
- Other ways I missed ... ?

Let me use `take(1000)` way first then. Let me know if you meant another way.

**Comment by rxin** [comment]
take 1001, and fail if there are 1001 records returned

**Comment by HyukjinKwon** [comment]
done!

**Comment by rxin** [comment]
ca we just take the first 1001 rows and call pandas' pivot and turn it back into a koalas dataframe? Only going through data once.

**Comment by HyukjinKwon** [comment]
yup.

**Comment by floscha** [comment]
So when wanting to process more than 1000 rows, the user would simply run `ks.DataFrame(kdf.to_pandas().T)` as a workaround, I guess? I can imagine cases where one would, for example, transpose a 2000x5 DataFrame (i.e. 10000 values) which could still be processed with pandas on a single machine. Should we maybe mention this somewhere in the docs/comments?

**Comment by rxin** [comment]
Yea - alternatively allow it to take a flag on max rows. I just worry about users waiting for a while and then everything crashes.

**Comment by floscha** [comment]
I think a flag similar to [options](https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html) in pandas sounds like a good idea, especially since pandas users should already be familiar with this. Wouldn't it then make sense to consequentially have the same for other methods like `to_pandas()` or `to_numpy()` as well, to guard users from shooting themselves in the foot?

**Comment by rxin** [comment]
Great idea. We might want different thresholds though because I can see
getting a million rows into a pandas DataFrame, but creating a million
columns will kill Sparks driver.

**Comment by HyukjinKwon** [comment]
> Should we maybe mention this somewhere in the docs/comments?

I think I mentioned this about how it works..

**Comment by floscha** [comment]
I think the comments are good for most parts. The only issue I see right now is that the `T` property cannot be called with the limit argument. So if a user gets the current `ValueError`, it might not be obvious to them that they must switch from `.T` to `.transpose(limit=n)`. Maybe you could thus change to error message to something like `... rows. Please use df.transpose(limit=<maximum number of rows>) to retrieve more than %s rows. Note that, ...` to give the user a better hint.

Also, if we wanted to add some pandas-like options, it would probably be best to move this (including further discussions) to another PR to keep this one clean, right?

File: databricks/koalas/frame.py [hunk 1]
```diff
@@ -1270,6 +1270,185 @@ def to_latex(self, buf=None, columns=None, col_space=None, header=True, index=Tr
         return validate_arguments_and_invoke_function(
             kdf.to_pandas(), self.to_latex, pd.DataFrame.to_latex, args)
 
+    # TODO: enable doctests once we drop Spark 2.3.x (due to type coercion logic
+    #  when creating arrays)
+    def transpose(self, limit=1000):
```
Comment (floscha) [inline databricks/koalas/frame.py]: Maybe you could add an `Optional[int]` type hint here.

File: databricks/koalas/frame.py [hunk 2]
```diff
@@ -1270,6 +1270,185 @@ def to_latex(self, buf=None, columns=None, col_space=None, header=True, index=Tr
...
+
+                >>> ks.DataFrame({'a': range(1001)}).transpose()  # doctest: +NORMALIZE_WHITESPACE
+                Traceback (most recent call last):
+                  ...
+                ValueError: Current DataFrame has more then the given limit 1000 rows. Note that,
+                before changing the given 'limit', this operation is considerably expensive.
+
+        Parameters
+        ----------
+        limit : int
```
Comment (floscha) [inline databricks/koalas/frame.py]: I guess you could also mention the default of 1000 here.

**Comment by HyukjinKwon** [comment]
Makes sense. Thanks for review @floscha

**Comment by HyukjinKwon** [comment]
I think I addressed all comments. Let me merge this one. let me know if you guys have more comments.

**Pull request merged**